### PR TITLE
RDBS-92 move string arrays to be unmanaged

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Lucene.Net/Search/FieldComparator.cs [hunk 1]
```diff
@@ -933,13 +936,13 @@ public override void  SetBottom(int bottom)
...
-		        get { return values[slot]; }
-		    }
+            {
+                get { return values[slot]; }
+            }
 
 		    public string[] GetValues()
 		    {
-		        return values;
+		        return values.ToStrings();
```
Comment (karmeli87) [inline src/Lucene.Net/Search/FieldComparator.cs]: As far as I can tell, no one use this... so I left it as is

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    dest._sortedSegments.Add(seg.Number, seg);
+            }
+
+            dest._strings[destPosition] = str;
+        }
+
+        public string[] ToStrings()
+        {
+            var strings = new string[_strings.Length];
+            for (int i = 1; i < _strings.Length; i++)
```
Comment (karmeli87) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: this should start from 0, and check for `null`s..

File: src/Lucene.Net/Search/FieldCacheImpl.cs [hunk 3]
```diff
@@ -782,42 +784,44 @@ public virtual StringIndex GetStringIndex(IndexReader reader, System.String fiel
...
-                {
-                    retArrayOrdered[i] = -1;
+                for (int i = 0; i < retArrayOrdered.Length; i++)
+                {
+                    retArrayOrdered[i] = -1;
                 }
-                System.String[] mterms = new System.String[reader.MaxDoc + 1];
+
+                var length = reader.MaxDoc + 1;
+                UnmanagedStringArray mterms = new UnmanagedStringArray(reader.MaxDoc + 1);
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCacheImpl.cs]:
```suggestion
                UnmanagedStringArray mterms = new UnmanagedStringArray(length);
```

File: src/Lucene.Net/Search/FieldCacheImpl.cs [hunk 4]
```diff
@@ -782,42 +784,44 @@ public virtual StringIndex GetStringIndex(IndexReader reader, System.String fiel
...
                 TermDocs termDocs = reader.TermDocs(state);
-                TermEnum termEnum = reader.Terms(new Term(field), state);
+                
+                SegmentTermEnum termEnum = (SegmentTermEnum)reader.Terms(new Term(field), state);
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCacheImpl.cs:808]: I'm pretty sure that we are guaranteed that this will always be the case here. 
I assume that all the tests are passing?

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs [hunk 5]
```diff
@@ -102,27 +102,54 @@ public override bool LessThan(FieldDoc docA, FieldDoc docB)
...
+                        {
+                            c = 1;
+                        }
+                        else if (fields.Array[i + fields.Offset].Locale == null)
+                        {
+                            c = us1.CompareTo(us2);
+                        }
+                        else
+                        {
+                            c = collators[i].Compare(us1.ToString(), us2.ToString());
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: Under what scenario do we have this?

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            public byte* CurrentPosition => Start + Used;
+            public int Free => Size - Used;
+            public int Used;
+
+            public Segment(int size)
+            {
+                Start = (byte*) Marshal.AllocHGlobal(size);
+                Used = 0;
+                Size = size;
+                Number = Interlocked.Increment(ref SegmentNumber);
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: Why is this required? This should just be a single threaded process, no?

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            ~Segment()
+            {
+                Dispose();
+            }
+        }
+
+        public struct UnmanagedString : IComparable
+        {
+            public byte* Start;
+            public long Owner;
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: Why do it this way? To avoid forcing a reference and the GC cost of that?

Comment (ayende) [reply src/Lucene.Net/Util/UnmanagedStringArray.cs]: I assume that this is for debug purposes only? If so, we want to drop this. It is actually too expensive for us (doubles the cost of this struct).

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                if (strA.IsNull && strB == null)
+                    return 0;
+
+                if (strB == null)
+                    return 1;
+
+                if (strA.IsNull)
+                    return -1;
+
+                Span<byte> managed = Encoding.UTF8.GetBytes(strB);
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: We shouldn't allocate a buffer here, instead, we want to use an already prepared buffer. Can use a thread local value, instead.

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                if (strB.IsNull)
+                    return 1;
+
+                if (strA.IsNull)
+                    return -1;
+
+                return strA.StringAsBytes.SequenceCompareTo(strB.StringAsBytes);
+            }
+
+            public static int CompareOrdinal(UnmanagedString strA, string strB)
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: Who is calling this method? If this is called a lot, it is better to turn that into a `span<byte> first to avoid the decoding work

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            fixed (char* s = str)
+            {
+                var size = (ushort) Encoding.UTF8.GetByteCount(s, str.Length);
+                var segment = GetSegment(size + sizeof(ushort));
+
+                segment.Add(size, out var position);
+                fixed (byte* r = new Span<byte>(position + sizeof(ushort), size))
+                {
+                    Encoding.UTF8.GetBytes(s, str.Length, r, size);
+                }
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]:
```suggestion
                Encoding.UTF8.GetBytes(s, str.Length, position + sizeof(ushort), size);
```

Comment (ayende) [reply src/Lucene.Net/Util/UnmanagedStringArray.cs]: why not like this?

File: src/Lucene.Net/Search/FieldComparator.cs [hunk 11]
```diff
@@ -778,24 +779,24 @@ public sealed class StringOrdValComparator:FieldComparator
 		{
 			
 			private int[] ords;
-			private System.String[] values;
+			private UnmanagedStringArray values;
 			private int[] readerGen;
 			
 			private int currentReaderGen = - 1;
-			private System.String[] lookup;
+			private UnmanagedStringArray lookup;
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldComparator.cs:786]: Should probably be: `UnmanagedStringArray.UnmanagedString[]` instead

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+                _strings[_index].Start = position;
+                _strings[_index].Owner = segment.Number;
+                _index++;
+            }
+        }
+
+        public UnmanagedString this[int position] => _strings[position];
+
+        public void CopyTo(UnmanagedStringArray dest, int destPosition, int position)
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: I don't follow why we need this and the `_sortedSegments` at all.
You are using to copy a value, but the lifetime of the `lookup` is the same as the `values`, so shouldn't be an issue there.

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            public readonly long Number;
+
+            public byte* Start;
+            public byte* CurrentPosition => Start + Used;
+            public int Free => Size - Used;
+            public int Used;
+
+            public Segment(int size)
+            {
+                Start = (byte*) Marshal.AllocHGlobal(size);
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: Here need to raise an event that RavenDB will wire into

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+                *(ushort*) CurrentPosition = size;
+
+                Used += sizeof(short) + size;
+            }
+
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+                if (Start != null)
+                    Marshal.FreeHGlobal((IntPtr) Start);
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: Here need to raise an event that RavenDB will wire into

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Lucene.Net/Search/FieldCacheImpl.cs (hunk 4 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:808]: yes

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs (hunk 5 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: Not sure TBH... this seems to be culture depended, so I left it as is..

File: src/Lucene.Net/Search/FieldComparator.cs (hunk 11 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldComparator.cs:786]: The problem is that the `lookup` is initially allocated more then needed (maxDoc + 1). 
So to pass it like `lookup = currentReaderValues.lookup` will require me to do an allocation of the actual used `UnamangedStrings`

File: src/Lucene.Net/Search/FieldCache.cs [hunk 15]
```diff
@@ -58,37 +61,65 @@ public virtual int BinarySearchLookup(System.String key)
             int low = 1;
             int high = lookup.Length - 1;
 
-            while (low <= high)
+            byte[] arr = null;
+            Span<byte> stringAsBytes = stackalloc byte[0]; // relax the compiler
+            var stringAsSpan = key.AsSpan();
+
+            var size = (ushort) Encoding.UTF8.GetByteCount(key);
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCache.cs]:
```suggestion
            var size = (ushort) Encoding.UTF8.GetByteCount(stringAsSpan );
```

File: src/Lucene.Net/Search/FieldComparator.cs (hunk 11 above)
Comment (ayende) [reply src/Lucene.Net/Search/FieldComparator.cs:786]: Why? You will copy the _reference_ to the array, not the reference itself.

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs (hunk 5 above)
Comment (ayende) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: Pretty sure that this isn't used at all in our usecase

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs [hunk 16]
```diff
@@ -102,27 +102,54 @@ public override bool LessThan(FieldDoc docA, FieldDoc docB)
...
+                        }
                     }
                     else
                     {
-                        c = collators[i].Compare(s1, s2);
+                        var s1 = (string) docA.fields[i];
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: Can one of them be of type ` UnmanagedStringArray.UnmanagedString` ?

File: src/Lucene.Net/Util/UnmanagedStringArray.cs [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        public UnmanagedString this[int position]
+        {
+            get => _strings[position];
+            set => _strings[position] = value;
+        }
+
+        public class MemoryMonitor
+        {
+            public static readonly MemoryMonitor Instance;
+
```
Comment (ayende) [inline src/Lucene.Net/Util/UnmanagedStringArray.cs]: Add Events here, so we can write this to our accounting.

**Review by ayende — APPROVED** [review]

File: src/Lucene.Net/Util/UnmanagedStringArray.cs (hunk 17 above)
Comment (karmeli87) [reply src/Lucene.Net/Util/UnmanagedStringArray.cs]: Would it be possible to access it simply by `UnmanagedStringArray.MemoryMonitor.Instance.AllocatedMemory` from raven? Or we want something more sophisticated?

File: src/Lucene.Net/Search/FieldComparator.cs (hunk 11 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldComparator.cs:786]: The problem is the `Length`, the `Length` of `UnmanagedStringArray` is based on how much we put in and not how much is allocated.

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs (hunk 5 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: So what should we do with that?

Comment (karmeli87) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: actually the `CompareTo` & `Compare` are culture specific... so I'm worried about the correctness here..
Is this intended to be culture-specific from the first place?

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs (hunk 16 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: Will unify both cases, just to be no the safe side

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs (hunk 5 above)
Comment (karmeli87) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: 3 tests failed when I throw in this code path.
`TestInternationalMultiSearcherSort`
`TestMultiSort`
`TestParallelMultiSort`

Comment (karmeli87) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: But I don't see any usages in ravendb itself for the `Locale`

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs [hunk 18]
```diff
@@ -102,54 +110,35 @@ public override bool LessThan(FieldDoc docA, FieldDoc docB)
...
+                    {
+                        if (s2 is UnmanagedStringArray.UnmanagedString us2)
                         {
-                            c = us1.CompareTo(us2);
+                            c = -us2.CompareTo(s1);
                         }
                         else
                         {
-                            c = collators[i].Compare(us1.ToString(), us2.ToString());
+                            c = s1.CompareTo(s2);
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: I don't understand what you are doing here.

File: src/Lucene.Net/Search/FieldDocSortedHitQueue.cs (hunk 5 above)
Comment (ayende) [reply src/Lucene.Net/Search/FieldDocSortedHitQueue.cs]: Yes, we don't care about this for Raven

**Review by ayende — APPROVED** [review]

**Review by ayende — APPROVED** [review]

File: src/Lucene.Net/Lucene.Net.csproj [hunk 19]
```diff
@@ -2,11 +2,19 @@
...
     <AssemblyTitle>Lucene.Net</AssemblyTitle>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
     <AssemblyName>Lucene.Net</AssemblyName>
     <PackageId>Lucene.Net</PackageId>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
   </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (ppekrol) [inline src/Lucene.Net/Lucene.Net.csproj]: Move that to the first property group block and remove this

File: src/Lucene.Net/Lucene.Net.csproj [hunk 20]
```diff
@@ -2,11 +2,19 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (ppekrol) [inline src/Lucene.Net/Lucene.Net.csproj]: Move that to the first property group block and remove this
